dlt/common/plugins/plugins.py

```diff
@@ -0,0 +1,55 @@
+from typing import Type
+from dlt.common.typing import TDataItem
+from dlt.common.pipeline import SupportsPipeline
+from dlt.common.configuration.specs import ContainerInjectableContext
+from dlt.common.schema.typing import TSchemaContract
+from dlt.common.configuration.specs import configspec
+
+
+class Plugin:
+    NAME: str = None
+
+    def __init__(self, pipeline: SupportsPipeline, step: str):
+        self.pipeline = pipeline
+        self.step = step
+
+    def on_step_start(self):
+        pass
+
+    def on_step_end(self):
+        pass
+
+    #
+    # extraction callbacks
+    #
+    def on_extractor_item_written(self, item: TDataItem, **kwargs):
+        pass
+
+    #
+    # normalizer callbacks
+    #
+    def filter_row(self, table_name: str, item: TDataItem, **kwargs):
+        return item
+    
+    #
+    # contracts callbacks
+    #
+    def on_schema_contract_violation(self, schema_contract: TSchemaContract, table_name: str, violating_item: TDataItem, **kwargs):
```

Comment:
We could repurpose the DataValidationError class which currently only is used for raising in the freeze case to send the specific violation info to this callback.